File: python/crypto_api/rsa/fulldomainhash_rsa_encryption.py
import hashlib

from rsa.rsa_data_classes import RSAKeyTuple
from rsa.rsa_encryption import RSAEncryption


class FullDomainHashRSAEncryption(RSAEncryption):

    @staticmethod
    def sign(secret_key: RSAKeyTuple, message: str) -> int:
        hash_method = hashlib.sha3_512
        h = int(hash_method(message.encode()).hexdigest(), 16)
        return pow(h, secret_key.key, secret_key.modul)

    @staticmethod
    def verify(public_key: RSAKeyTuple, message: str, signature: int) -> bool:
        hash_method = hashlib.sha3_512
        h = int(hash_method(message.encode()).hexdigest(), 16)
        return h % public_key.key == pow(signature, public_key.key, public_key.modul)

File: java/cryptoapi/rsa/FulldomainHashRSAEncryption.java
package cryptoapi.rsa;

import cryptoapi.math_lib.MathMisc;

import java.math.BigInteger;

public class FulldomainHashRSAEncryption extends RSAEncryption {


    public static BigInteger sign(RSAKeyTuple keyTuple, BigInteger m) {
        return new BigInteger(MathMisc.hash(m)).modPow(keyTuple.key, keyTuple.modul);
    }

    public static boolean verify(RSAKeyTuple publicKey, BigInteger m, BigInteger sigma) {
        BigInteger h = new BigInteger(MathMisc.hash(m));
        return h.mod(publicKey.modul).equals(sigma.modPow(publicKey.key, publicKey.modul));
    }
}
